daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/layers/api/Layer.scala

```diff
@@ -0,0 +1,275 @@
+ * so that they can be found and dynamically loaded.
+ *
+ * The SPI creates an instance the class of which is used as a factory to create the
+ * instances actually used by the Daffodil runtime. Compilation of the static information about
+ * the layer occurs only once and is then shared by all runtime instances.
+ *
+ * Instances of derived layer classes can be stateful. They are private to threads, and each time a layer
+ * is encountered during parse/unparse, an instance is created for that situation.
+ *
+ * All the static information about the layer is provided in the arguments.
+ *
+ * The rest of the Layer class implements the
+ * layer decode/encode logic, which is done as part of deriving one's Layer class from the
+ * Layer base class.
+ *
+ * About variables: Layer logic may read and write variables. Variables being read are parameters to
+ * the layer algorithm. Variables being written are outputs (such as checksums) from the layer algorithm.
+ * Variables being written must be undefined, since variables in DFDL are single-assignment.
+ * Variables being read must be defined before being read by the layer, and this is true for both
+ * parsing and unparsing. When unparsing, variables being read cannot be forward-referencing to parts
+ * of the DFDL infoset that have not yet been unparsed.
+ *
+ * @param layerName the name that will appear in the DFDL schema to identify the layer
+ * @param supportedLayerLengthKinds list of the layer length kinds the layer supports
+ * @param supportedLayerLengthUnits list of the layer length units the layer supports
+ * @param isRequiredLayerEncoding true if the layer is textual and so needs the layerEncoding property
+ * @param optLayerVariables a LayerVariables structure describing the variables the layer accesses
+ */
+abstract class Layer(
+  val layerName: String,
+  val supportedLayerLengthKinds: java.util.List[JLayerLengthKind],
+  val supportedLayerLengthUnits: java.util.List[JLayerLengthUnits],
+  val isRequiredLayerEncoding: Boolean,
+  val optLayerVariables: Optional[LayerVariables],
+) {
+
+  final def name(): String =
+    layerName // name() method with empty args is required by SPI loader
+
+  /**
+   * Called exactly once when the schema is compiled to do extra checking that the layer is being used properly.
+   * The thrown exception becomes a SchemaDefinitionError at schema compile time.
+   *
+   * Example checks are:
+   * - layerEncoding is constant and is a single-byte charset
+   * - layerLength, if constant, is within a maximum value range
+   * - layerBoundaryMark string, if constant, is not too long and contains only allowed characters.
+   * These things can be required to be constant by this check, or it can check their values for legality
+   * if they happen to be constant. Since these are runtime-valued properties (can be expressions), then if the
+   * layer allowed that, they must also be checked at runtime.
```

Comment:
In thinking more about how things might work if limiting and transform layers were all the same thing, we need to consider how layerLength, layerBoundaryMark, etc. settings are provided to layers. I guess these would have to be variables? For example, a boundary mark might look like:

```xml
<sequence>
  <annotation>
    <appinfo source="http://www.ogf.org/dfdl/">
      <dfdl:newVariableInstance ref="bm:boundaryMark" defaultValue="\r\n" />
    </appinfo>
  </annotation>
  <sequence dfdlx:layer="boundaryMark">
     <sequence dfdlx:layer="lineFolded_AIS" >
       ...
     </sequence>
  </sequence>
  ...
</sequence>
```

It's a bit more verbose, but it's much more general. I guess the idea is there are no other `dfdlx:layer*` properties, there is only `dfdlx:layer`. Layers can be either transform or limiting (or both), and all configuration of those layers must be done via variables.

Reply:
This is what I will try next. More verbose, but that's mostly XSD's fault.  Only one thing to learn: variables.